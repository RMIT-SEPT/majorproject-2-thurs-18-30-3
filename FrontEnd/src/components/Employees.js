import React, {useEffect, useState} from 'react'
import {makeStyles} from '@material-ui/core/styles'
import Snackbar from '@material-ui/core/Snackbar'
import Alert from '@material-ui/lab/Alert'
import MaterialTable from 'material-table'
import Axios from 'axios'

import '../containers/App.css'

function Employees(props) {
  const classes = useStyles()
  const [columns] = useState([
    {title: 'Username', field: 'uname'},
    {title: 'Name', field: 'name'},
    {title: 'Email', field: 'email'},
    {title: 'Mobile', field: 'mobile'},
  ])
  const [employees, setEmployees] = useState([])
  const [alertMsg, setAlertMsg] = useState('')
  const [alertErrorMsg, setAlertErrorMsg] = useState('')

  useEffect(() => {
    if (props.employees.length > 0) {
      setEmployees(props.employees)
    }
  }, [props.employees])

  async function onRowAdd(newData) {
    // TODO: call real api
    try {
      await Axios.post('https://5f50f63c5e98480016123379.mockapi.io/employees', newData)
      setEmployees([...employees, newData])
      setAlertMsg(`${newData.name} has been added`)
    } catch ({messsage}) {
      setAlertErrorMsg(messsage)
    }
  }

  async function onRowUpdate(newData, oldData) {
    try {
      await Axios.put(`https://5f50f63c5e98480016123379.mockapi.io/employees/${oldData.id}`, newData)
      setEmployees([...employees, newData])
    } catch ({messsage}) {
      setAlertErrorMsg(messsage)
    }

    setTimeout(() => {
      const data = [...employees]
      data[data.indexOf(oldData)] = newData
      setEmployees(data)
    }, 600)
  }

  async function onRowDelete(oldData) {
    try {
      await Axios.delete(`https://5f50f63c5e98480016123379.mockapi.io/employees/${oldData.id}`)
      const data = [...employees]
      data.splice(data.indexOf(oldData), 1)
      setEmployees(data)
    } catch ({messsage}) {
      setAlertErrorMsg(messsage)
    }
  }

  return (
    <div className={classes.root}>
      <Snackbar
        anchorOrigin={{vertical: 'top', horizontal: 'right'}}
        open={alertMsg !== ''}
        onClose={() => setAlertMsg('')}>
        <Alert onClose={() => setAlertMsg('')} severity="success">
          {alertMsg}
        </Alert>
      </Snackbar>
      <Snackbar
        anchorOrigin={{vertical: 'top', horizontal: 'right'}}
        open={alertErrorMsg !== ''}
        onClose={() => setAlertErrorMsg('')}>
        <Alert onClose={() => setAlertErrorMsg('')} severity="error">
          {alertErrorMsg}
        </Alert>
      </Snackbar>
      <MaterialTable
        title="Employees"
        columns={columns}
        data={employees}
        editable={{
          onRowAdd,
          onRowUpdate,
          onRowDelete,
        }}
      />
    </div>
  )
}

const useStyles = makeStyles((theme) => ({
  root: {
    margin: theme.spacing(5),
  },
  table: {
    minWidth: 650,
  },
}))

export default Employees
