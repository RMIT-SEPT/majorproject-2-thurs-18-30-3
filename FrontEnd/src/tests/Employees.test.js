import React from 'react'

import Employees from '../components/Employees'
// import renderer from 'react-test-renderer'
import {render} from '@testing-library/react'

beforeEach(() => {})

describe('Employees test', () => {
  const data = [
    {id: '2', name: 'Phoebe Swaniawski', mobile: 'mobile 2', email: 'Una.Hoppe0@gmail.com', uname: 'phoebe.swan'},
  ]
  const mockFunc = jest.fn()

  // Waiting for a new release of Material Table
  /*it('renders employees components successfully', () => {
    const employeesComponent = renderer
      .create(
        <Employees
          employees={data}
          onRowAdd={mockFunc}
          onRowUpdate={mockFunc}
          onRowDelete={mockFunc}
          onAlertClose={mockFunc}
          alertMsg=""
          alertErrorMsg=""
        />
      )
      .toJSON()
    expect(employeesComponent).toMatchSnapshot()
  })*/

  it('shows success alert message', () => {
    const alertMsg = 'success message'

    const {getByText} = render(
      <Employees
        employees={data}
        onRowAdd={mockFunc}
        onRowUpdate={mockFunc}
        onRowDelete={mockFunc}
        onAlertClose={mockFunc}
        alertMsg={alertMsg}
        alertErrorMsg=""
      />
    )
    expect(getByText(alertMsg)).toBeTruthy()
  })

  it('shows error alert message', () => {
    const alertErrorMsg = 'error alert message asdfasdfasdfasdf'

    const {getByText} = render(
      <Employees
        employees={data}
        onRowAdd={mockFunc}
        onRowUpdate={mockFunc}
        onRowDelete={mockFunc}
        onAlertClose={mockFunc}
        alertMsg=""
        alertErrorMsg={alertErrorMsg}
      />
    )
    expect(getByText(alertErrorMsg)).toBeTruthy()
  })
})
