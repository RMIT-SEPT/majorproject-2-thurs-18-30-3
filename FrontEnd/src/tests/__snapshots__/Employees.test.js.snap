// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employees test renders employees components successfully 1`] = `
<div
  className="makeStyles-root-1"
/>
`;
